### PR TITLE
[PodSecurity] Add annotations denoting the exemption reason and the enforcement policy used

### LINKED ISSUE #103923: [PodSecurity] Extra audit annotations
The current alpha implementation doesn't record all the audit annotations that are defined in the KEP
(https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/2579-psp-replacement/README.md#audit-annotations):

1. [ ] `pod-security.kubernetes.io/enforce-policy = <policy_level>:<resolved_version>` Record which policy was evaluated
   for enforcing mode.
    - ~~Resolved version is the actual version of the policy that was evaluated, so in the case of
      `latest` or future versions, it will be `latest@<version>` where `<version>` is the tagged
      version of the apiserver or webhook (e.g. `latest@v1.22.5-build.8`).~~
    - This annotation is only recorded when a policy is enforced. Specifically, it will not be
      recorded for irrelevant updates or exempt requests.
2. [x] ~~`pod-security.kubernetes.io/audit-policy = <policy_level>:<resolved_version>` Same as `enforce-policy`, but for
   audit mode policies (only included when an audit policy is set).~~
3. [x] ~~`pod-security.kubernetes.io/enforce-violations = <policy violations>` When an enforcing policy is violated, record
   the violations here.~~ _Already captured in response status message._
4. [x] `pod-security.kubernetes.io/audit-violations = <policy violations>` When an audit mode policy is violated, record
   the violations here.
    - [ ] Actual implementation sets `pod-security.kubernetes.io/audit = <policy violations>`. I think we should switch to `audit-violations` to distinguish from the `audit` policy level label.
5. [ ] `pod-security.kubernetes.io/exempt = [user, namespace, runtimeClass]` For exempt requests, record the parameters
   that triggered the exemption here.
6. [x] _(not defined in the KEP)_ `pod-security.kubernetes.io/error = <policy parse error>` When the policy label isn't parsed correctly
    - If we want to keep this, we should probably set it for all errors.

We need to clean up the annotations, decide which of these we want, and update the KEP to match the intent.

### PR DESCRIPTION
**Pull request by stlaz**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR adds a couple of annotations for the PodSecurity admission. These annotations are supposed to bring more clarity into how the pod was being evaluated. It also changes the `audit` annotation on audit violations to `audit-violation` to distinguish it from the namespace annotation used to mark the audit policy level.

#### Which issue(s) this PR fixes:
Fixes #103923

#### Does this PR introduce a user-facing change?

```release-note
The pods and pod controllers that are exempted from the PodSecurity admission process are now marked with the "pod-security.kubernetes.io/exempt: user/namespace/runtimeClass" annotation, based on what caused the exemption.

The enforcement level that allowed or denied pod during PodSecurity admission is now marked by the "pod-security.kubernetes.io/enforce-policy" annotation.

The annotation that informs about audit policy violations changed from ""pod-security.kubernetes.io/audit" to ""pod-security.kubernetes.io/audit-violation".
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
[KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/2579-psp-replacement/README.md
```

### CODE REVIEW

**Review by tallclair — COMMENTED** [review]
Thanks! Left a few small suggestions.

File: staging/src/k8s.io/pod-security-admission/admission/admission.go [hunk 1]
```diff
@@ -427,7 +437,7 @@ func (a *Admission) EvaluatePod(ctx context.Context, nsPolicy api.Policy, nsPoli
 
 	// TODO: reuse previous evaluation if audit level+version is the same as enforce level+version
 	if result := policy.AggregateCheckResults(a.Evaluator.EvaluatePod(nsPolicy.Audit, podMetadata, podSpec)); !result.Allowed {
-		auditAnnotations["audit"] = fmt.Sprintf(
+		auditAnnotations["audit-violation"] = fmt.Sprintf(
```
[the remark this answers is not in the document]
Comment (stlaz) [reply staging/src/k8s.io/pod-security-admission/admission/admission.go:440]: I think I might have just overlooked the plural, this should be fixed now

File: staging/src/k8s.io/pod-security-admission/admission/admission.go [hunk 2]
```diff
@@ -308,8 +308,12 @@ func (a *Admission) ValidatePod(ctx context.Context, attrs api.Attributes) *admi
...
 	}
 	// short-circuit on exempt namespaces and users
-	if a.exemptNamespace(attrs.GetNamespace()) || a.exemptUser(attrs.GetUserName()) {
-		return sharedAllowedResponse()
+	if a.exemptNamespace(attrs.GetNamespace()) {
+		return allowedByExemptResponse("namespace")
+	}
+
+	if a.exemptUser(attrs.GetUserName()) {
+		return allowedByExemptResponse("user")
```
[the remark this answers is not in the document]
Comment (stlaz) [reply staging/src/k8s.io/pod-security-admission/admission/admission.go:316]: I used the same pattern as the allowed response is using, then

File: staging/src/k8s.io/pod-security-admission/api/constants.go [hunk 3]
```diff
@@ -43,4 +43,8 @@ const (
 	AuditVersionLabel   = labelPrefix + "audit-version"
 	WarnLevelLabel      = labelPrefix + "warn"
 	WarnVersionLabel    = labelPrefix + "warn-version"
+
+	ExemptionReasonLabel = labelPrefix + "exempt"
+	AuditViolationLabel  = labelPrefix + "audit-violation"
+	EnforcedPolicyLabel  = labelPrefix + "enforce-policy"
```
[the remark this answers is not in the document]
Comment (stlaz) [reply staging/src/k8s.io/pod-security-admission/api/constants.go:49]: Thanks for the suggestion, it should be incorporated in the PR now.

**Review by tallclair — APPROVED** [review]
/lgtm
/approve

**Pull request merged**